pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAcksTracker.java

```diff
@@ -30,15 +30,17 @@
 import java.util.concurrent.TimeUnit;
 import org.apache.pulsar.client.api.Message;
 import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.MessageIdAdv;
 import org.apache.pulsar.client.api.RedeliveryBackoff;
 import org.apache.pulsar.client.impl.conf.ConsumerConfigurationData;
+import org.apache.pulsar.common.util.collections.ConcurrentLongLongPairHashMap;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 class NegativeAcksTracker implements Closeable {
     private static final Logger log = LoggerFactory.getLogger(NegativeAcksTracker.class);
 
-    private HashMap<MessageId, Long> nackedMessages = null;
+    private ConcurrentLongLongPairHashMap nackedMessages = null;
```

Comment:
If it's still proved to have much smaller memory usage, we can go ahead with `ConcurrentLongLongPairHashMap` for now and write our own non-concurrent hash map for long pairs in future

Reply:
- It seems that `partitionIndex` field is unnecessary, so we can use `HashMap<Pair<Long, Long>, Long>` too, mapping `(ledgerId, entryId)` to `timestamp`, I will supplement the test for this option.
- The size of memory consumption is not related to the detailed content of the field, each long field take up 8 byte, which is an invariable.

Reply:
@thetumbled @BewareMyPower  In the Pulsar code base, we are already using https://fastutil.di.unimi.it/ . It contains memory optimized data structures. I introduced this dependency quite recently in PIP-379 changes. 
Since this is replacing a HashMap, I'd assume that a non-thread safe implementation from Fastutil would be sufficient.

Reply:
example usages of Fastutil:
* https://github.com/apache/pulsar/blob/137df29f85798b00de75460a1acb91c7bc25453f/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/DrainingHashesTracker.java#L44
* https://github.com/apache/pulsar/blob/137df29f85798b00de75460a1acb91c7bc25453f/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PendingAcksMap.java#L101

Reply:
FastUtil contains also primitive pair and tuple implementations. In the above example of PendingAcksMap, `IntIntPair` is used.

Reply:
No, ConcurrentLongLongPairHashMap is thread safety.

Reply:
Could you provide a FastUtil class to map Pair<Long, Long> to Pair<Long, Long>? @lhotari

Reply:
> Could you provide a FastUtil class to map Pair<Long, Long> to Pair<Long, Long>? @lhotari 
> 

There's no specialized map implementation, however there are primitive pair classes such as LongLongPair. That could be used with an ordinary hashmap and it would avoid the overhead of boxing.

Reply:
Yes, `LongLongPair` can be used to replace `Pair<Long, Long>`.

```java
public class LongLongImmutablePair implements LongLongPair, Serializable {
    private static final long serialVersionUID = 0L;
    protected final long left;
    protected final long right;
```

Reply:
Hi, i have update the latest experiment result in the pr, which shows that the `ConcurrentLongLongPairHashMap` is still the best choice for us to store large amount of the entries.